File: C#/Visitor.cs
using System;
using System.Collections.Generic;

interface IVisitor {
    void Visit(ElementA element);
    void Visit(ElementB element);
}

class ConcreteVisitor : IVisitor {
    public void Visit(ElementA element) {
        Console.WriteLine("ConcreteVisitor is visiting ElementA");
    }

    public void Visit(ElementB element) {
        Console.WriteLine("ConcreteVisitor is visiting ElementB");
    }
}

interface IElement {
    void Accept(IVisitor visitor);
}

class ElementA : IElement {
    public void Accept(IVisitor visitor) {
        visitor.Visit(this);
    }
}

class ElementB : IElement {
    public void Accept(IVisitor visitor) {
        visitor.Visit(this);
    }
}

class ObjectStructure {
    private readonly List<IElement> elements = new List<IElement>();

    public void Attach(IElement element) {
        elements.Add(element);
    }

    public void Detach(IElement element) {
        elements.Remove(element);
    }

    public void Accept(IVisitor visitor) {
        foreach (var element in elements) {
            element.Accept(visitor);
        }
    }
}

class Program {
    static void Main(string[] args) {
        ObjectStructure structure = new ObjectStructure();
        structure.Attach(new ElementA());
        structure.Attach(new ElementB());

        IVisitor visitor = new ConcreteVisitor();
        structure.Accept(visitor);
    }
}

File: Dart/Visitor.dart
abstract class Visitor {
  void visitElementA(ElementA elementA);
  void visitElementB(ElementB elementB);
}

abstract class Element {
  void accept(Visitor visitor);
}

class ElementA implements Element {
  final String _name;
  ElementA(this._name);
  String get name => _name;
  @override
  void accept(Visitor visitor) {
    visitor.visitElementA(this);
  }
}

class ElementB implements Element {
  final int _value;
  ElementB(this._value);
  int get value => _value;
  @override
  void accept(Visitor visitor) {
    visitor.visitElementB(this);
  }
}

class ConcreteVisitor implements Visitor {
  @override
  void visitElementA(ElementA elementA) {
    print('Visitando ElementA com nome ${elementA.name}');
  }
  @override
  void visitElementB(ElementB elementB) {
    print('Visitando ElementB com valor ${elementB.value}');
  }
}

void main() {
  final elements = <Element>[
    ElementA('ElementA'),
    ElementB(10),
    ElementA('Outro ElementA'),
    ElementB(20),
  ];

  final visitor = ConcreteVisitor();

  elements.forEach((element) {
    element.accept(visitor);
  });
}
